assets/js/base/context/cart-checkout/payment-methods/use-payment-method-registration.js

```diff
@@ -54,64 +54,81 @@ const usePaymentMethodRegistration = (
 ) => {
 	const [ isInitialized, setIsInitialized ] = useState( false );
 	const { isEditor } = useEditorContext();
-	const { shippingAddress } = useShippingDataContext();
+	const {
+		selectedRates: selectedShippingMethods,
+		shippingAddress,
+	} = useShippingDataContext();
 	const { cartTotals, cartNeedsShipping } = useStoreCart();
 	const canPayArgument = useRef( {
 		cartTotals,
 		cartNeedsShipping,
 		shippingAddress,
+		selectedShippingMethods,
 	} );
 
 	useEffect( () => {
 		canPayArgument.current = {
 			cartTotals,
 			cartNeedsShipping,
 			shippingAddress,
+			selectedShippingMethods,
 		};
-	}, [ cartTotals, cartNeedsShipping, shippingAddress ] );
+	}, [
+		cartTotals,
+		cartNeedsShipping,
+		shippingAddress,
+		selectedShippingMethods,
+	] );
 
-	const resolveCanMakePayments = useCallback( async () => {
-		let initializedPaymentMethods = {},
-			canPay;
-		const setInitializedPaymentMethods = ( paymentMethod ) => {
-			initializedPaymentMethods = {
-				...initializedPaymentMethods,
+	const refreshCanMakePayments = useCallback( async () => {
+		let availablePaymentMethods = {};
+		const addAvailablePaymentMethod = ( paymentMethod ) => {
+			availablePaymentMethods = {
+				...availablePaymentMethods,
 				[ paymentMethod.name ]: paymentMethod,
 			};
 		};
 		for ( const paymentMethodName in registeredPaymentMethods ) {
-			const current = registeredPaymentMethods[ paymentMethodName ];
+			const paymentMethod = registeredPaymentMethods[ paymentMethodName ];
 
+			// In editor, shortcut so all payment methods show as available.
 			if ( isEditor ) {
-				setInitializedPaymentMethods( current );
+				addAvailablePaymentMethod( paymentMethod );
 				continue;
 			}
 
+			// In front end, ask payment method if it should be available.
 			try {
-				canPay = await Promise.resolve(
-					current.canMakePayment( canPayArgument.current )
+				const canPay = await Promise.resolve(
+					paymentMethod.canMakePayment( canPayArgument.current )
 				);
 				if ( canPay ) {
 					if ( canPay.error ) {
 						throw new Error( canPay.error.message );
 					}
-					setInitializedPaymentMethods( current );
+					addAvailablePaymentMethod( paymentMethod );
 				}
 			} catch ( e ) {
+				// If user is admin, show payment `canMakePayment` errors as a notice.
 				handleRegistrationError( e );
```

Comment:
This notice feels tightly coupled to Stripe integration - though maybe this could be useful for other methods in future. 

I'm thinking we could decouple the initialisation of the payment methods from `canMakePayment`. Not sure if we need to be involved in the init explicitly - maybe payment methods can manage this themselves. 

Related, is it essential that we wrap `canMakePayment` in try/catch? Feels heavy-handed, though I could understand that might be useful if this was an "initialise yourself" callback.

Reply:
> This notice feels tightly coupled to Stripe integration - though maybe this could be useful for other methods in future.

How so? It will only happen if the payment method throws an error from it's `canMakePayment` callback. There's no reference to stripe in this logic.

> I'm thinking we could decouple the initialisation of the payment methods from canMakePayment. Not sure if we need to be involved in the init explicitly - maybe payment methods can manage this themselves.

I agree with splitting up logic more, however I disagree with your assessment. Checkout IS responsible for initializing payment methods, otherwise payment methods have no idea _when_ to initialize themselves. To be clear, in this existing logic (not with your changes) the purpose of this hook is to:

- grab all the registered payment methods.
- run through everyone of them and fire the `canMakePayment` hook to determine whether checkout should include that payment method in it's "ready to make payment" state (which essentially determines whether the payment displays or not).
- this is a checkout initialization routine for payment methods to determine whether they display on _initial_ checkout load. Whether or not payment methods do any of their initialization in `canMakePayment` is something checkout doesn't care about. It only cares that at this point, payment methods are saying - "yup you can show me to the customer".

I have some other thoughts on this in a later comment but I do think we need to keep this _checkout initialization state for payments_ distinct from where `canMakePayment` is called AFTER this for any potential updates. The reason is that with these changes, payment methods will need to know this state because `canPayment` is no longer called only once. So while checkout doesn't care if payment methods do initialization when `canMakePayment` is called, payment methods typically _might_ need to do their initialization during this call to determine the answer to "Can I make payment?" and they might only want to do that once. So basically we need to provide data to the payment method to make that decision.

> Related, is it essential that we wrap canMakePayment in try/catch? Feels heavy-handed, though I could understand that might be useful if this was an "initialise yourself" callback.

What about this feels heavy-handed? It's unclear how that term applies to using try/catch? 

This is essential because of the possibility third party payment methods might cause an error here and we want to catch that early and handle as opposed to having the third party payment method break checkout. The reason why there is a `CURRENT_USER_IS_ADMIN` check inside `handleRegistrationError` is so that we only surface the payment method related error to the store admin when they go to figure out why the payment method isn't showing up for a shopper (which is critical because we have no idea what the thrown error might contain from the third party payment method and I don't think we want to risk exposing sensitive information).

Reply:
> How so? It will only happen if the payment method throws an error from it's canMakePayment callback. There's no reference to stripe in this logic.

Stripe is the only payment method right now that's doing anything initialisation in `canMakePayment`. But it's quite possible this pattern will work well for other payment gateways too.

> I agree with splitting up logic more, however I disagree with your assessment. Checkout IS responsible for initializing payment methods, otherwise payment methods have no idea when to initialize themselves.

For sure, I understand how it currently works. I'm wondering if we need to do this inside `canMakePayment` or if we can have a separate hook for initialisation. Coupling init to `canMakePayment` makes the interface more complicated – would be nice if we could avoid that.

What do you think about the possibility of making the init callback a separate property, and focus `canMakePayment` just on availability? Interested to hear your thoughts on that approach. 

Anyway I'll take another pass on this and revise it in line with your feedback :)

Reply:
> So basically we need to provide data to the payment method to make that decision.

One thought on this - do we need to manage that boolean for the gateways? They can easily detect when they haven't initialised themselves. Just thinking out loud here, looking for ways to simplify the API where we can.

Reply:
> One thought on this - do we need to manage that boolean for the gateways?

It's a signal that checkout has done the initial loading of payment methods. You're right payment methods could likely manage their own internal needs here. I'm happy with you trying my suggested approach without passing this through to the payment method. But I still think the new hook will need the value.

Reply:
> Stripe is the only payment method right now that's doing anything initialisation in canMakePayment. But it's quite possible this pattern will work well for other payment gateways too.

To be clear, this is _explicitly_ defensive code to catch ANY errors from payment methods borking something in their `canMakePayment` callback. For critical flows like checkout I think it's crucial we have this kind of protection. So yes, Stripe is the only payment method right now that's doing any initialization, but that is _definitely_ not going to remain that way.

Reply:
> For sure, I understand how it currently works. I'm wondering if we need to do this inside canMakePayment or if we can have a separate hook for initialisation. Coupling init to canMakePayment makes the interface more complicated – would be nice if we could avoid that.

Let's think of this from the perspective of checkout and the payment method.

Checkout:

- only cares if the payment method reports it's ready to make payments.
- It doesn't care how the payment method initializes.
- It doesn't care how the payment method figures out it's ready to make payments.
- It just cares that if `canMakePayment` returns true, then the registered payment method is included for display to the shopper in the app.

Payment method

- doesn't know when checkout is going to fire the callback it provides on `canMakePayment` and really doesn't care. 
- it just cares that the values it's provided by checkout are enough for it to report whether it can make payment or not.

Given that information, I'm not sure we need to provide a separate interface for initializing  when essentially that's totally up to the payment method and how it chooses/needs to initialize. In writing this, I realize this further supports [your comment](https://github.com/woocommerce/woocommerce-gutenberg-products-block/pull/2840#discussion_r452491572) that in fact we don't need to send along the `isInitialized` boolean to the payment method provided callback.

Essentially the major change here from the perspective of payment methods is that `canMakePayment` can be invoked more than once and so payment methods need to account for that if there's anything expensive they do in the callback.

Reply:
That works! I'll implement the optimisation suggested by @Aljullu and then rebase #2831 to make sure we've solved the problem for COD :)